### PR TITLE
[MDS] Adds datasource filter for version decoupling

### PR DESCRIPTION
**Pull request by DarshitChanpura**
### Description
This change adds a datasource filter function to allow for version decoupling for MDS feature.

### Category
Enhancement

### Why these changes are required?
To allow for version decoupling in MDS.

### What is the old behavior before changes and new behavior after changes?
Before: Datasources with all versions were shown.
After: Datasource with only compatible versions are shown.

### Issues Resolved
- Resolves https://github.com/opensearch-project/security-dashboards-plugin/issues/1971

### Testing
Unit tests

### Check List
- [x] New functionality includes testing
~- [ ] New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch_dashboards.json [hunk 1]
```diff
@@ -15,5 +15,7 @@
     "dataSourceManagement"
   ],
   "server": true,
-  "ui": true
+  "ui": true,
+  "supportedOSDataSourceVersions": ">=1.0.0",
```
Comment (derek-ho) [inline opensearch_dashboards.json:19]: I think we only tested with 1.3 + should we changes this to 1.3?

Comment (DarshitChanpura) [reply opensearch_dashboards.json:19]: @ZilongX Can you help answer this?

File: public/utils/datasource-utils.ts [hunk 2]
```diff
@@ -12,9 +12,12 @@
  *   express or implied. See the License for the specific language governing
  *   permissions and limitations under the License.
  */
-
+import semver from 'semver';
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply public/utils/datasource-utils.ts:15]: not directly, but it is transitively imported from core: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/c8496f83bf92f8a8230bc0ddbed218a5480dd49b/package.json#L239

Also imported as a dependency of cypress and few other dependencies. Checkout yarn.lock.

File: opensearch_dashboards.json (hunk 1 above)
Comment (ZilongX) [reply opensearch_dashboards.json:19]: 1.3 check is for the managed side, for opensource side let's take in all supported OS versions

File: public/utils/datasource-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (derek-ho) [reply public/utils/datasource-utils.ts:15]: We should explicitly add it as a dependency if we need it. If core removes it someday or cypress removes it one day then we will be broken. We should protect ourselves from that.

**Review by derek-ho — CHANGES_REQUESTED** [review]
See comment: https://github.com/opensearch-project/security-dashboards-plugin/pull/2051/files#r1688545956

**Review by ZilongX — APPROVED** [review]

File: public/utils/datasource-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply public/utils/datasource-utils.ts:15]: added as a local dependency to unblock this PR.

**Comment by DarshitChanpura** [comment]
> See comment: https://github.com/opensearch-project/security-dashboards-plugin/pull/2051/files#r1688545956

addressed.

**Review by derek-ho — APPROVED** [review]
Does the yarn.lock need to be updated?

**Review by scrawfor99 — APPROVED** [review]

**Review by cwperks — APPROVED** [review]

**Review by derek-ho — DISMISSED** [review]
Lets take a look at the cypress failures for MDS to make sure there are no regressions prior to merging

**Review by scrawfor99 — APPROVED** [review]

**Review by cwperks — DISMISSED** [review]

**Review by RyanL1997 — APPROVED** [review]

**Review by cwperks — APPROVED** [review]

**Pull request merged**